library/core/src/array/mod.rs

```diff
@@ -364,6 +364,28 @@ impl<T: Ord, const N: usize> Ord for [T; N] {
     }
 }
 
+#[cfg(not(bootstrap))]
+#[stable(feature = "copy_clone_array_lib", since = "1.54.0")]
+impl<T: Copy, const N: usize> Copy for [T; N] {}
+
+#[cfg(not(bootstrap))]
+#[stable(feature = "copy_clone_array_lib", since = "1.54.0")]
+impl<T: Clone, const N: usize> Clone for [T; N] {
+    fn clone(&self) -> Self {
+        let mut other: [MaybeUninit<T>; N] = MaybeUninit::uninit_array();
+        for i in 0..N {
+            // SAFETY: `get_unchecked(i)` must use a valid index `i`.
+            // `self` and `other` are length N, and 0 <= i < N
+            unsafe {
+                other.get_unchecked_mut(i).write(self.get_unchecked(i).clone());
+            }
+        }
+        // SAFETY: All elements in `other` must be initialized.
+        // `other` is length N, and the prior loop iterates N times.
+        unsafe { MaybeUninit::array_assume_init(other) }
```

Comment:
```suggestion
        unsafe { collect_into_array_unchecked(&mut self.iter().cloned()) }
```

Using `collect_into_array_unchecked` would reduce code duplication and enhance handling of an eventual panic in `T::clone`.

Reply:
Yes I realized from one of the test failures that I can't get away with not having some kind of `Guard`-like thing going on. However I'm pretty sure I can't just call `collect_into_array` because I have to actually call `clone` on each element, right? I was planning on factoring out `Guard` into something that both this and `collect_into_array` can use.

Reply:
> because I have to actually call `clone` on each element

[`Iterator::cloned`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.cloned) is an iterator adapter that clones each element, so it should be fine.

Reply:
Could also just use `self.each_ref().map(Clone::clone)`, I think, which is safe and should be effectively the same.

(It just definitely shouldn't be one that adds another implementation of an array builder guard type.)

Reply:
> Could also just use self.each_ref().map(Clone::clone), I think, which is safe and should be effectively the same.

Unfortunately this is not a good solution because LLVM is very bad at eliding unnecessary copies of temporary arrays.  

Example: https://godbolt.org/z/foYb6fr4E

Reply:
@a1phyr Wow, that's even worse than the https://github.com/rust-lang/rust/issues/79914 I was expecting.